// ngx-extended-pdf-viewer/esm2015/lib/toolbar/pdf-zoom-toolbar/pdf-zoom-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class PdfZoomToolbarComponent {
    constructor() {
        this.zoomChange = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    emitZoomChange(event) {
        debugger;
        this.zoomChange.emit(event);
    }
}
PdfZoomToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'pdf-zoom-toolbar',
                template: "<div [class.invisible]=\"!showZoomButtons\" id=\"toolbarViewerMiddle\">\n  <div class=\"splitToolbarButton\">\n    <button type=\"button\"\n            id=\"zoomOut\"\n            class=\"toolbarButton zoomOut\"\n            title=\"Zoom Out\"\n            tabindex=\"21\"\n            data-l10n-id=\"zoom_out\">\n              <span data-l10n-id=\"zoom_out_label\">Zoom Out</span>\n    </button>\n    <div class=\"splitToolbarButtonSeparator\"></div>\n    <button type=\"button\"\n            id=\"zoomIn\"\n            class=\"toolbarButton zoomIn\"\n            title=\"Zoom In\"\n            tabindex=\"22\"\n            data-l10n-id=\"zoom_in\">\n      <span data-l10n-id=\"zoom_in_label\">Zoom In</span>\n    </button>\n  </div>\n  <pdf-zoom-dropdown\n      style=\"width: unset; max-width: unset;padding-top:6px\"\n      (zoomChange)=\"emitZoomChange($event)\"></pdf-zoom-dropdown>\n</div>\n",
                styles: [""]
            }] }
];
PdfZoomToolbarComponent.propDecorators = {
    showZoomButtons: [{ type: Input }],
    zoomChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PdfZoomToolbarComponent.prototype.showZoomButtons;
    /** @type {?} */
    PdfZoomToolbarComponent.prototype.zoomChange;
}
