// ngx-extended-pdf-viewer/esm2015/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of ngx-extended-pdf-viewer
 */
export { NgxExtendedPdfViewerComponent } from './lib/ngx-extended-pdf-viewer.component';
export { NgxExtendedPdfViewerService } from './lib/ngx-extended-pdf-viewer.service';
export { NgxExtendedPdfViewerModule } from './lib/ngx-extended-pdf-viewer.module';
export { defaultOptions } from './lib/options/default-options';
export { FindState } from './lib/events/find-result';
export {} from './lib/events/pages-loaded-event';
export {} from './lib/events/page-number-change';
export {} from './lib/events/page-rendered-event';
export {} from './lib/events/pages-loaded-event';
export {} from './lib/events/pages-rotation-event';
export {} from './lib/events/pdf-downloaded-event';
export {} from './lib/events/pdf-loaded-event';
export {} from './lib/events/scale-changing-event';
export {} from './lib/events/sidebarview-changed';
export { VerbosityLevel } from './lib/options/verbosity-level';
