// ngx-extended-pdf-viewer/esm5/lib/toolbar/pdf-findbar/pdf-findbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef } from '@angular/core';
var PdfFindbarComponent = /** @class */ (function () {
    function PdfFindbarComponent() {
    }
    PdfFindbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'pdf-findbar',
                    template: "<div\n  class=\"findbar hidden doorHanger\"\n  [class.invisible]=\"!showFindButton\"\n  id=\"findbar\"\n  [style.transform]=\"'scale('+mobileFriendlyZoomScale+')'\"\n  [style.transformOrigin]=\"'left center'\"\n  [style.left]=\"findbarLeft\"\n  [style.top]=\"findbarTop\">\n    <pdf-find-input-area></pdf-find-input-area>\n    <pdf-findbar-options-one-container></pdf-findbar-options-one-container>\n    <pdf-findbar-options-two-container></pdf-findbar-options-two-container>\n    <pdf-findbar-options-three-container></pdf-findbar-options-three-container>\n    <pdf-findbar-message-container></pdf-findbar-message-container>\n</div>\n",
                    styles: [""]
                }] }
    ];
    PdfFindbarComponent.propDecorators = {
        showFindButton: [{ type: Input }],
        mobileFriendlyZoomScale: [{ type: Input }],
        findbarLeft: [{ type: Input }],
        findbarTop: [{ type: Input }],
        findbarInputArea: [{ type: Input }]
    };
    return PdfFindbarComponent;
}());
export { PdfFindbarComponent };
if (false) {
    /** @type {?} */
    PdfFindbarComponent.prototype.showFindButton;
    /** @type {?} */
    PdfFindbarComponent.prototype.mobileFriendlyZoomScale;
    /** @type {?} */
    PdfFindbarComponent.prototype.findbarLeft;
    /** @type {?} */
    PdfFindbarComponent.prototype.findbarTop;
    /** @type {?} */
    PdfFindbarComponent.prototype.findbarInputArea;
}
