// @ng-select/ng-select/esm2015/ng-select/ng-select.types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function NgOption() { }
/* TODO: handle strange member:
[name: string]: any;
*/
/** @type {?|undefined} */
NgOption.prototype.index;
/** @type {?|undefined} */
NgOption.prototype.htmlId;
/** @type {?|undefined} */
NgOption.prototype.mapped;
/** @type {?|undefined} */
NgOption.prototype.selected;
/** @type {?|undefined} */
NgOption.prototype.disabled;
/** @type {?|undefined} */
NgOption.prototype.marked;
/** @type {?|undefined} */
NgOption.prototype.label;
/** @type {?|undefined} */
NgOption.prototype.value;
/** @type {?|undefined} */
NgOption.prototype.parent;
/** @type {?|undefined} */
NgOption.prototype.children;
/** @enum {number} */
const KeyCode = {
    Tab: 9,
    Enter: 13,
    Esc: 27,
    Space: 32,
    ArrowUp: 38,
    ArrowDown: 40,
    Backspace: 8,
};
export { KeyCode };
KeyCode[KeyCode.Tab] = 'Tab';
KeyCode[KeyCode.Enter] = 'Enter';
KeyCode[KeyCode.Esc] = 'Esc';
KeyCode[KeyCode.Space] = 'Space';
KeyCode[KeyCode.ArrowUp] = 'ArrowUp';
KeyCode[KeyCode.ArrowDown] = 'ArrowDown';
KeyCode[KeyCode.Backspace] = 'Backspace';
